### PR TITLE
Enable persistent expert loops from coarse-tile hints

### LINKED ISSUE #3928: Explicit expert-axis coarse tiling cannot form a persistent expert loop
## Summary

An explicit `spyre_hint` over the expert dimension cannot currently realize a dense expert FFN as one counted device loop.

For:

- `x: [T,H]`, shared by every expert
- gate/up weights: `[E,H,F]`
- down weights: `[E,F,H]`
- routing weights: `[E,T,1]`

the intended physical execution is:

1. stage `x` once;
2. execute one expert body in an `E`-trip device loop;
3. advance only the three weight operands and routing operand;
4. accumulate each weighted result into one `[T,H]` value in LX;
5. write that value once after the loop.

## Campaign status

Resolved and accepted. Every supporting mechanism and the final carried accumulator are in `main` through #4042 (`3358f39e`). The emitted structure, device numerics, and same-stack device performance have passed acceptance.

## Original compiler gaps

The implementation campaign identified four gaps:

1. When tiling reduces `E` to extent 1, dependency extraction removes the corresponding symbol before the original expert-to-expert address step is preserved.
2. Coarse tiling inserts temporary read copies for expert weights even when the matmul could read the advancing HBM slice directly.
3. A terminal reduction is represented as an ordinary tiled reduction rather than one loop-carried accumulator.
4. Scratchpad lifetime analysis sees the loop body only once and therefore cannot represent values that remain live across every loop iteration.

## Focused reproductions

The snippets below use the existing named-dimension test helpers to associate `E`, `T`, `H`, and `F` with tensor axes.

### 1. Preserve the expert address step and keep `x` fixed

```python
E, T, H, F = 2, 64, 64, 64
x = torch.randn(T, H, dtype=torch.float16)
w_backing = torch.randn(H, E, F, dtype=torch.float16)
w = w_backing.permute(1, 0, 2)  # [E,H,F]

with spyre_hint(num_tiles_per_dim={"E": E}):
    y = torch.matmul(x.unsqueeze(0), w)  # [E,T,F]
```

For the counted `E` loop, the generated operands must satisfy:

```text
x address step = 0
w address step = one expert weight slab
```

Tiling `E` to extent 1 must not erase the second fact.

### 2. Remove a temporary weight copy only after proving direct access

Use the same matmul above and inspect the post-layout IR/generated source.

Expected:

- the matmul reads the original graph-input weight argument;
- that argument carries the nonzero per-expert address step;
- no temporary read-copy buffer feeds the matmul.

If the original weight layout is incompatible with the matmul, the copy must remain. Copy removal is therefore a proven optimization, not an unconditional rewrite.

### 3. Carry one reduction value through the loop

```python
E, T, H = 2, 64, 64
values = torch.randn(E, T, H, dtype=torch.float16)

with spyre_hint(
    num_tiles_per_dim={"E": E},
    work_div={"T": 32},
):
    result = values.sum(dim=0)  # [T,H]
```

Expected generated structure:

```text
fill [T,H] accumulator once
for e in range(E):
    accumulator += values[e]
drain accumulator to result once
```

The per-tile local reduction has extent 1, so it contributes one value; the loop-carried add performs the real cross-expert reduction.

### 4. Preserve scratchpad lifetime across repeated loop trips

Allocator-level reproduction:

```python
persistent = LifetimeBoundBuffer(
    "persistent",
    size=64,
    uses=[0, 1],
    lifetime_end_override=5,
)
later = LifetimeBoundBuffer("later", size=64, uses=[2, 3])
```

Although the textual use lists do not overlap, `persistent` is live through the counted loop. The allocator must therefore assign the two buffers different addresses. The Python and native layout solvers must agree.

### 5. Require stable row ownership

Compile the reduction reproduction twice:

```python
# No ownership contract: retain the ordinary reduction path.
spyre_hint(num_tiles_per_dim={"E": E})

# Stable row ownership: loop-carried accumulation is legal.
spyre_hint(num_tiles_per_dim={"E": E}, work_div={"T": 32})
```

Only the second form may use one persistent accumulator. Every loop trip must assign a given token row to the same core.

## Expected result

For a hinted dense expert FFN, generated code should contain:

- one counted expert loop;
- one copy of the expert body;
- one invariant `x` staging operation;
- advancing gate, up, down, and routing operands;
- one non-aliasing `[T,H]` LX accumulator;
- no expert-expanded activation in HBM;
- one output drain after the loop.

Graphs without the explicit hint must retain their existing behavior.

## Landed implementation

- Capture input address advances before range division and use them only when a hinted dimension is later squeezed out.
- Preserve read copies through layout propagation, then remove a copy only after proving direct access is layout-compatible.
- Represent a terminal flat expert sum as fill, per-iteration add, and one post-loop drain.
- Extend the general scratchpad lifetime representation through counted loops.
- Require explicit stable row ownership before enabling the carried accumulator.

## Reduced complete FFN reproduction

This final case proves that the focused behaviors compose.

```python
E, T, H, F = 2, 64, 64, 64

# x: [T,H]
# gate_w/up_w: [E,H,F]
# down_w: [E,F,H]
# route: [E,T,1]

with spyre_hint(
    num_tiles_per_dim={"E": E},
    work_div={"T": 32},
):
    gate = torch.matmul(x.unsqueeze(0), gate_w)
    up = torch.matmul(x.unsqueeze(0), up_w)
    hidden = torch.nn.functional.gelu(gate, approximate="tanh") * up
    down = torch.matmul(hidden, down_w)
    result = (down * route).sum(dim=0)  # [T,H]
```

Acceptance requires device correctness against the ordinary PyTorch expression and the complete generated-program structure listed above.

The corresponding end-to-end regression in #3892 is `TestCoarseTileMoEBroadcastMatmulE2E.test_hinted_dense_expert_loop_keeps_one_body`.

## Accepted performance

These measurements cover two different boundaries:

- Expert FFN only, chunked versus persistent: `373.1 ms → 42.6 ms` (`8.8×`).
- Full Gemma 4 26B prefill, batch size 1 and sequence length 512: `6.4 s → 1.6 s` (`4.0×`).

The first isolates the expert FFN; the second is the end-to-end model prefill result.

## Tracking

Part of #3565.

All focused compiler issues and PRs are complete:

- #4035 → #4039: counted-loop scratchpad lifetime (merged);
- #4036 → #4040: operand address steps (merged);
- #4096: order-preserving work-division name translation (merged);
- #4075: allocation-only empty classification (merged);
- #4078: invariant activation staging (merged);
- #4037 → #4041: proven direct weight reads (merged);
- #4038 → #4042: carried terminal sum (merged).

Final compiler state:

```text
main@3358f39e contains the complete accepted dense expert-loop path
```

Structural, numerical-correctness, and same-stack device-performance acceptance passed. Prerequisite #3890 has merged into `main`.

### PR DESCRIPTION
**Pull request by AdnanHoque**
Fixes #3928
Part of #3565
Builds on the coarse-tiling size-one fixes merged in #3890

Related but not fixed:

- #3888 — broadcast matmul ambiguity for more than one expert per tile
- #3891 — eager-result host normalization performance regression

## Campaign status

Superseded by the focused split PRs #4075, #4039, #4040, #4096, #4078, #4041, and #4042. Every piece is now merged into `main` through `3358f39e`, and structural, numerical-correctness, and same-stack device-performance acceptance passed.

## Accepted performance

These measurements cover two different boundaries:

- Expert FFN only, chunked versus persistent: `373.1 ms → 42.6 ms` (`8.8×`).
- Full Gemma 4 26B prefill, batch size 1 and sequence length 512: `6.4 s → 1.6 s` (`4.0×`).

The first isolates the expert FFN; the second is the end-to-end model prefill result.

## What this does

Makes an explicit coarse-tiling hint sufficient to express the dense all-expert execution schedule:

- emit one expert body inside one counted device loop;
- stage the shared `[T,H]` input once in LX;
- advance the gate, up, down, and routing operands to the next expert each trip;
- keep the `[T,H]` running sum in LX through the loop and drain it once;
- remove temporary weight-copy operations only after layout propagation proves that the matmul can read the advancing HBM slice directly.

This reuses the existing hint, coarse-tiling, counted-loop, layout, work-division, scratchpad, and SuperDSC paths. It adds no model-facing operation or separate planner.

## Compiler gaps addressed

| Problem | Existing behavior | Fix | Compiler area |
|---|---|---|---|
| Lost expert address step | Tiling `E` to size 1 removes the symbol before the next-expert address step is reconstructed. | Preserve the input's original step before tiling and use it when the dimension disappears. | `coarse_tile.py`, `loop_info.py` |
| Temporary weight buffers | Coarse tiling stages each expert weight slice through a temporary buffer. | Remove the copy only after layout propagation proves the matmul can read the advancing HBM slice directly. | `coarse_tile.py`, `propagate_layouts.py` |
| No loop-carried sum | The terminal expert sum is represented as an ordinary tiled reduction. | Create one `[T,H]` accumulator, add one contribution per expert, and drain it once after the loop. | `coarse_tile.py` |
| Loop lifetime ends early | Scratchpad planning sees one textual loop body and can end `x` or accumulator lifetime before all trips finish. | Extend their allocation lifetime through the counted loop in both layout solvers. | `scratchpad/*`, `perm_layout_native.cpp` |
| Unstable row ownership | A running sum is unsafe when successive loop trips can assign rows to different cores. | Require explicit row work division before enabling the carried accumulator. | `coarse_tile.py`, work-division metadata |

## Scope

The new behavior is restricted to loops created by an explicit `spyre_hint`. Automatic span-overflow tiling keeps its existing behavior. If direct weight streaming cannot be proven safe, the existing copy remains.

The loop-carried reduction path accepts only a terminal flat `sum` with no other consumer. Ordinary and nested reductions keep the existing path.

The three commits contain 781 production additions and 107 test additions on current main. Most of the production code is the three physical requirements that hints did not previously express: preserving an address step after a tiled dimension becomes size one, carrying a value's LX lifetime across repeated loop iterations, and draining a terminal reduction once after the loop.

## Validation

- Reduced complete expert FFN (`E=2, T=64, H=64, F=64`) passes device correctness against CPU.
- Generated source has one loop body, four advancing operands, no activation `hbm_pool`, no in-loop HBM restickify, and one reduction drain.
- #3890's `E=128` shared-LHS broadcast-matmul correctness test passes on device.
- Focused compiler suites: 416 passed, 5 skipped. One pre-existing four-layer nested-launch case reached an unsupported DeepTools bundle opcode on the available pod; its one-layer form passed.
- After rebasing onto current main, Ruff, Python syntax, C++ formatting, commit signatures, DCO trailers, and `git diff --check` pass. Fresh CI is pending.

Performance effects from eager result-layout normalization are tracked separately in #3891. This PR does not use a latency threshold as a correctness gate.

## Current CI status

The PR head is based directly on current main (06cee473). All coarse-tiling and focused persistent-loop checks pass. The remaining failed device jobs are the same three broad suites failing on current main—matmul, LX reduction scalar, and reduction scalar—so they are not attributed to this PR.

### CODE REVIEW

**Comment by ani300** [comment]
## Model-load applier for the MoE dense-hint path: `dma_moe_expert_weight_to_spyre`

Building the Gemma-4 MoE adapter (hf-adapters, `google/gemma-4-26B-A4B-it`) on top of this coarse-tile persistent-expert hint path, we needed a **model-load-time applier** that lays each expert weight out in a single device layout usable by *both* MoE FFN formulations:

- **prefill (persistent hint-body)** — the weight is a matmul operand `A[m,k] @ B[k,n]`, so it must be **sticked on its free dim `n`**;
- **decode (loop-on-topk gather)** — `weight[idx]` is an indirect access, so the **indexed (expert) dim must be outermost**.

One layout satisfies both: expert dim `E` outermost, free dim `F` split into sticks → device dims `[E, C, F // eps, eps]`, contraction dim `C` un-split in the middle. Logical PyTorch shape/stride is unchanged; only the device layout carries the arrangement. This retired a dual-weight-set materialization (~85 GiB on a ~96 GiB card → OOM) down to one ~42.5 GiB set.

**Validated prefill-only on-card** (gemma-4-26B-A4B, 512-token prefill): load 77.9s with a single weight set and **no `0x340f` OOM**, cold gen 163.8s → `' the'`, warm gen 6.8s → `' the'`. Device layout confirmed `device_size=[128, 2816, 11, 64]` (E outermost, 64-elem stick at fp16).

It's additive-only (one new function in `model_utils.py`, Python-only, no existing code touched, no C++ rebuild). Mirrors the existing `_dma_to_spyre_indirect_access` 2-D template. Sharing here in case it's useful to fold into the MoE story on this branch — happy to open a separate PR if you'd prefer.

Diff (against `model_utils.py`):

```diff
@@ -199,6 +199,66 @@ def _dma_to_spyre_indirect_access(
     return dst
 
 
+def dma_moe_expert_weight_to_spyre(
+    weight: torch.Tensor,
+    target_dtype: torch.dtype | None = None,
+) -> torch.Tensor | None:
+    """Transfer a rank-3 ``[E, C, F]`` MoE expert weight to Spyre.
+
+    The expert dim ``E`` is placed outermost on the device and the free/output
+    dim ``F`` is split into stick-sized blocks, giving device dims
+    ``[E, C, F // eps, eps]`` where ``eps`` is the elements-per-stick for the
+    device dtype. This one layout serves both MoE paths:
+
+      * the decode gather ``weight[idx]`` needs the indexed (expert) dim
+        outermost -- the "indirect access" requirement; and
+      * the matmul weight operand ``A[m,k] @ B[k,n]`` needs the weight sticked
+        on its free dim ``n`` (here ``F``), with the contraction dim ``C``
+        un-split in the middle.
+
+    ``gate``/``up`` are logical ``[E, H, M]`` (contract H, free M); ``down`` is
+    logical ``[E, M, H]`` (contract M, free H). The logical PyTorch
+    shape/stride is unchanged; only the device layout carries this arrangement.
+
+    Uses the 3-arg device-dims ``SpyreTensorLayout`` overload with the *device*
+    dtype (``get_device_dtype``).
+
+    Requires ``F % eps == 0``; otherwise the sticks can't tile the free dim, so
+    we warn and return ``None`` to signal the caller to fall back to a default
+    move (which still loads and runs, just without the shared-layout
+    optimization).
+
+    Caller must ensure ``weight.ndim == 3``.
+    """
+    assert weight.ndim == 3, "MoE expert-weight path is for rank-3 [E,C,F] only"
+
+    if not weight.is_contiguous():
+        weight = weight.contiguous()
+    dev_dtype = target_dtype if target_dtype is not None else weight.dtype
+
+    experts, contract, free = weight.shape
+    # elems_per_stick is dtype-aware (64 at fp16/bf16, 32 at fp32), so query it
+    # rather than hardcoding a stick size.
+    eps = SpyreTensorLayout(list(weight.shape), dev_dtype).elems_per_stick()
+    if free % eps != 0:
+        warnings.warn(
+            f"MoE expert-weight free dim {free} is not a multiple of the Spyre "
+            f"stick size {eps} for dtype {dev_dtype}; falling back to the "
+            "default layout (no shared-layout optimization) for this weight.",
+            stacklevel=2,
+        )
+        return None
+
+    layout = SpyreTensorLayout(
+        [experts, contract, free // eps, eps],  # device_size: E outermost
+        [contract * free, free, eps, 1],  # stride_map
+        get_device_dtype(dev_dtype),
+    )
+    dst = spyre_empty_with_layout(weight.size(), weight.stride(), dev_dtype, layout)
+    copy_tensor(weight, dst, non_blocking=False)
+    return dst
+
+
 # --- Model loading ---------------------------------------------------
```

**Comment by AdnanHoque** [comment]
Superseded by four smaller draft PRs, each tied to one focused reproduction from #3928:

- #4039 — counted-loop scratchpad lifetime
- #4040 — preserve the operand step after size-one tiling
- #4041 — proven-safe direct expert-weight reads
- #4042 — loop-carried terminal sum and stable row ownership

Dependency order: #4039 → #4040 → #4041 → #4042. The final stacked tree is byte-for-byte identical to this PR head.

For clarity, this PR's current red checks are infrastructure failures, not failing compiler tests: both wheel jobs stopped in `astral-sh/setup-uv` while downloading uv 0.12.6 with HTTP 429, before build or test execution.
